iroh-bytes/src/get/progress.rs

```diff
@@ -0,0 +1,206 @@
+                .children
+                .entry(id)
+                .or_insert_with(|| BlobState::new(hash)),
+        }
+    }
+    fn get_blob_mut(&mut self, blob_id: &BlobId) -> Option<&mut BlobState> {
+        match blob_id {
+            BlobId::Root => Some(&mut self.root),
+            BlobId::Child(id) => self.children.get_mut(id),
+        }
+    }
+
+    fn get_by_progress_id(&mut self, progress_id: ProgressId) -> Option<&mut BlobState> {
+        let blob_id = *self.progress_id_to_blob.get(&progress_id)?;
+        self.get_blob_mut(&blob_id)
+    }
+
+    /// Update the state with a new [`DownloadProgress`] event for this transfer.
+    pub fn on_progress(&mut self, event: DownloadProgress) {
+        match event {
+            DownloadProgress::InitialState(s) => {
+                *self = s;
+            }
+            DownloadProgress::FoundLocal {
+                child,
+                hash,
+                size,
+                valid_ranges,
+            } => {
+                let blob = self.get_or_insert_blob(BlobId::from_child_id(child), hash);
+                blob.size = Some(size);
+                blob.local_ranges = Some(valid_ranges);
+            }
+            DownloadProgress::Connected => self.connected = true,
+            DownloadProgress::Found {
+                id: progress_id,
+                child,
+                hash,
+                size,
+            } => {
+                let blob_id = BlobId::from_child_id(child);
+                let blob = self.get_or_insert_blob(blob_id, hash);
+                if blob.size.is_none() {
+                    blob.size = Some(BaoBlobSize::Verified(size));
+                }
+                blob.progress = BlobProgress::Progressing(0);
+                self.progress_id_to_blob.insert(progress_id, blob_id);
+                self.current = Some(blob_id);
+            }
+            DownloadProgress::FoundHashSeq { hash, children } => {
+                if hash == self.root.hash {
+                    self.root.child_count = Some(children);
+                } else {
+                    // I think it is an invariant of the protocol that `FoundHashSeq` is only
+                    // triggered for the root hash.
+                }
+            }
+            DownloadProgress::Progress { id, offset } => {
+                if let Some(blob) = self.get_by_progress_id(id) {
+                    blob.progress = BlobProgress::Progressing(offset);
+                }
+            }
+            DownloadProgress::Done { id } => {
+                if let Some(blob) = self.get_by_progress_id(id) {
+                    blob.progress = BlobProgress::Done;
+                    self.progress_id_to_blob.remove(&id);
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+/// The id of a blob in a transfer
+#[derive(
+    Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, std::hash::Hash, Serialize, Deserialize,
+)]
+pub enum BlobId {
```

Comment:
Nice, might be worth moving this further into the get part of iroh-bytes for clarity.

Reply:
Moved it up!